modules/generator/registry/counter.go

```diff
@@ -0,0 +1,140 @@
+package registry
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"go.uber.org/atomic"
+)
+
+type counter struct {
+	name   string
+	labels []string
+
+	// seriesMtx is used to sync modifications to the map, not to the data in series
+	seriesMtx sync.RWMutex
+	series    map[uint64]*counterSeries
+
+	onAddSeries    func(count uint32) bool
+	onRemoveSeries func(count uint32)
+}
+
+type counterSeries struct {
+	// labelValues should not be modified after creation
+	labelValues []string
+	value       *atomic.Float64
+	lastUpdated *atomic.Int64
+}
+
+var _ Counter = (*counter)(nil)
+var _ metric = (*counter)(nil)
+
+func newCounter(name string, labels []string) *counter {
+	return &counter{
+		name:   name,
+		labels: labels,
+		series: make(map[uint64]*counterSeries),
+	}
+}
+
+func (c *counter) setCallbacks(onAddSeries func(uint32) bool, onRemoveSeries func(count uint32)) {
```

Comment:
Any reason not to pass these functions in the constructor? I didn't see these fields changed other than when the metric is created.

Reply:
I think I did this so I could make `setCallbacks` part of the `metric` interface. It doesn't save much code though and setting it in the `newCounter`/`newHistogram` function is just as easy.